trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -110,31 +115,220 @@ func newNodeBufferList(
+			wpBlocks:        wpBlocks,
+			rsevMdNum:       rsevMdNum,
+			dlInMd:          dlInMd,
+			limit:           limit,
+			base:            base,
+			head:            ele,
+			tail:            ele,
+			count:           1,
+			persistID:       rawdb.ReadPersistentStateID(db),
+			stopCh:          make(chan struct{}),
+			waitStopCh:      make(chan struct{}),
+			forceKeepCh:     make(chan struct{}),
+			waitForceKeepCh: make(chan struct{}),
+			keepFunc:        keepFunc,
+		}
+	}
+
+	go nf.loop()
+
+	log.Info("new node buffer list", "proposed block interval", nf.wpBlocks,
+		"reserve multi difflayers", nf.rsevMdNum, "difflayers in multidifflayer", nf.dlInMd,
+		"limit", common.StorageSize(limit), "layers", layers, "persist id", nf.persistID, "base_size", size)
+	return nf, nil
+}
+
+// newNodeBufferListForRecovery initializes the node buffer list for recovering diff layers
+func newNodeBufferListForRecovery(db ethdb.Database, freezer *rawdb.ResettableFreezer, base *multiDifflayer,
+	limit, wpBlocks, rsevMdNum, dlInMd uint64) (*nodebufferlist, error) {
+	nbl := &nodebufferlist{
 		db:              db,
 		wpBlocks:        wpBlocks,
 		rsevMdNum:       rsevMdNum,
 		dlInMd:          dlInMd,
 		limit:           limit,
 		base:            base,
-		head:            ele,
-		tail:            ele,
-		count:           1,
 		persistID:       rawdb.ReadPersistentStateID(db),
 		stopCh:          make(chan struct{}),
 		waitStopCh:      make(chan struct{}),
 		forceKeepCh:     make(chan struct{}),
 		waitForceKeepCh: make(chan struct{}),
-		keepFunc:        keepFunc,
 	}
+	freezerLength, err := freezer.Ancients()
+	if err != nil {
+		log.Error("Failed to get freezer ancients", "error", err)
+		return nil, err
+	}
+	tail, err := freezer.Tail()
+	if err != nil {
+		log.Error("Failed to get freezer tail", "error", err)
+		return nil, err
+	}
+	log.Info("Ancient db meta info", "persistent_state_id", nbl.persistID, "freezer_length", freezerLength,
+		"freezer_tail", tail, "waiting_recover_num", freezerLength-nbl.persistID)
 
-	go nf.loop()
+	var (
+		wg           sync.WaitGroup
+		start        = time.Now()
+		startStateID = nbl.persistID + 1
+	)
+	startBlock, err := readBlockNumber(freezer, startStateID)
+	if err != nil {
+		log.Error("Failed to read start block number", "error", err, "state_id", startStateID)
+		return nil, err
+	}
+	endBlock, err := readBlockNumber(freezer, freezerLength)
+	if err != nil {
+		log.Error("Failed to read end block number", "error", err, "state_id", freezerLength)
+		return nil, err
+	}
+	blockIntervals := nbl.createBlockInterval(startBlock, endBlock)
+	stateIntervals, err := nbl.createStateInterval(freezer, startStateID, freezerLength, blockIntervals)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("block intervals info", "blockIntervals", blockIntervals, "stateIntervals", stateIntervals, "startBlock",
+		startBlock, "endBlock", endBlock)
+
+	nbl.linkMultiDiffLayers(len(blockIntervals))
+	errChan := make(chan error, 1)
```

Comment:
When multiple errors occur, both the sender and the receiver will be blocked.

Reply:
Fixed

Reply:
@sysvm hi, I think this issue still exists. Can you take another look? When there are 2 or more errors, a goroutine gets stuck at sending the error, and the main thread gets stuck at wg.Wait().

Reply:
Fixed